data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/reconciler/impl/ResourcesReconcilerImpl.java

```diff
@@ -50,137 +60,150 @@ private ResourcesReconcilerImpl(
-            );
-          } else if (oldResource.hasIngress() && !newResource.hasIngress()) {
-            futures.add(
-              this.ingressReconcilerListener.onDeleteIngress(oldResource, oldResource.getIngress())
-            );
-          } else if (newResource.hasIngress()) {
-            futures.add(
-              this.ingressReconcilerListener.onUpdateIngress(newResource, newResource.getIngress())
-            );
-          }
-        }
-
-        // Reconcile egress
-        if (isReconcilingEgress()) {
-          final var oldEgressesKeys = oldResource
-            .getEgressesList()
-            .stream()
-            .map(DataPlaneContract.Egress::getUid)
-            .collect(Collectors.toSet());
-          final var newEgresses = newResource
-            .getEgressesList()
-            .stream()
-            .collect(Collectors.toMap(DataPlaneContract.Egress::getUid, Function.identity()));
-
-          final var diff = CollectionsUtils.diff(oldEgressesKeys, newEgresses.keySet());
-
-          // Handle added
-          for (String uid : diff.getAdded()) {
-            final var newEgress = newEgresses.get(uid);
-            this.cachedEgresses.put(uid, newEgress);
-            futures.add(
-              this.egressReconcilerListener.onNewEgress(newResource, newEgress)
-            );
-          }
-
-          // Handle removed
-          for (String uid : diff.getRemoved()) {
-            futures.add(
-              this.egressReconcilerListener.onDeleteEgress(oldResource, this.cachedEgresses.remove(uid))
-            );
-          }
-
-          // Handle intersection
-          for (String uid : diff.getIntersection()) {
-            final var newEgress = newEgresses.get(uid);
-            final var oldEgress = this.cachedEgresses.replace(uid, newEgress);
-
-            if (!egressEquals(newEgress, oldEgress) || !resourceAreEquals) {
-              futures.add(
-                this.egressReconcilerListener.onUpdateEgress(newResource, newEgress)
-              );
-            }
-          }
-        }
-      }
 
-      if (newResource == null) {
-        // Resource deleted
-        resourcesIterator.remove();
-
-        // Reconcile ingress
-        if (isReconcilingIngress() && oldResource.hasIngress()) {
-          futures.add(
-            this.ingressReconcilerListener.onDeleteIngress(oldResource, oldResource.getIngress())
-          );
-        }
-
-        // Reconcile egress
-        if (isReconcilingEgress() && oldResource.getEgressesCount() != 0) {
-          for (DataPlaneContract.Egress egress : oldResource.getEgressesList()) {
-            this.cachedEgresses.remove(egress.getUid());
-            futures.add(
-              this.egressReconcilerListener.onDeleteEgress(oldResource, egress)
-            );
-          }
-        }
+    final List<Future> futures = new ArrayList<>(newResourcesMap.size() + this.cachedResources.size());
```

Comment:
More than creating an array with futures and fill it, in each stream you can end with a map and then join the streams at the end:

```
final var diffStream = diff.getRemoved().stream().....map(...)
[...]

CompositeFuture.all(Stream.concat(Stream.concat(diffStream, addedStream), intersectionStream)).mapEmpty();
```

Reply:
Yeah, I used your solution first, then I went for this because we can preallocate all possible futures we want and a stream has no size concept, thus creating less garbage.

Reply:
Ok sounds good!